### ISSUE TITLE
Containers started on convergence that installs docker are stopped due to delayed restart

### ISSUE THREAD

**Issue by agperson** [chef-cookbooks/docker#793]
### Cookbook version
2.13.5

### Platform Details
CentOS 7 / systemd

### Scenario:
On first convergence, the docker_service resource queues a delayed restart, which stops all running containers.  This was addressed in #695 by making the restart immediate, but reverted in #701 due to a restart loop.

**Comment by chasebolt** [comment]
i run into this issue as well now. will need to take some time to address the restart loop properly so we can go back to restarting docker_service immediately.

**Comment by agperson** [comment]
I was trying to get a handle on what's happening to submit a PR, but got stuck in my own little loop -- it seems like the logic is that the docker_service resource's :start action is both doing configuration/installing and then starting the service, and then calling *itself* for a restart?

**Comment by agperson** [comment]
I tried simply moving the service management resource into the restart action, which seemed fine.  But I'm unable to duplicate the previous error locally to verify the fix.  Gonna leave this to someone more familiar with the architecture of this cookbook.

**Comment by chasebolt** [comment]
this is actually causing kitchen failures now. going to look deeper into this soon.

**Comment by someara** [comment]
This smells like a bug in Chef. This definitely used to Do The Right Thing.

I've experimented with changing the timings to :immediate, but it still seems to schedule it in the parent run_context.

I've added a couple more resources to the auto.rb recipe in the test cookbook to make this easy to experiment with. As a spot check, I tried changing the notification timings to :immediate inside the systemd provider. It yields the same behavior: a restart at the end chef-client run.

https://github.com/chef-cookbooks/docker/blob/master/test/cookbooks/docker_test/recipes/auto.rb

The expected behavior is that the `docker_service` resource is fully converged before the `docker_image` and `docker_container` resources start.

-s

**Comment by someara** [comment]
@jkeiser @thommay @lamont-granquist @tas50 @cheeseplus 
Has this behavior changed some time in the past year?

-s

**Comment by someara** [comment]
The notification should be happening at the private run_context of the Custom Resource (ala use_inline_resources), not at the end of the parent run_context or main chef-client run.

**Comment by lamont-granquist** [comment]
- notifications and finding resources now "bubbles up" and will find resources in outer run contexts
- when it finds the resource it will schedule a delayed notification into run_context of the target resource it finds

if you're trying to write non-delayed immediate accumulators you may need to follow this pattern now:

```ruby
    t = begin
          run_context.resource_collection.find_local(:multipackage_internal => "collected packages #{action}")
        rescue Chef::Exceptions::ResourceNotFound
          multipackage_internal "collected packages #{action}" do
            package_name Array.new
            version Array.new
            action action
          end
        end
```

see https://github.com/lamont-cookbooks/multipackage/blob/b82cf462e9d382c1a73af6be2e6a92ac82399625/definitions/default.rb

**Comment by someara** [comment]
Ug. Really? 

That's completely non-intuitive, dangerous as hell because it violates the closure (you have no control over what other cookbooks put into the resource collection), and it Makes Chef Hard.

What version did that change in?

**Comment by someara** [comment]
Removing restarts completely here feffc9ba0fab52377ac208ce1c504d89c8945626
Releasing fix in 2.13.9

**Comment by lamont-granquist** [comment]
It changed months (6+?) back.  It enables the more common pattern where someone wants to add a service to the outer run context in a recipe and then send a notification to that resource from within a provider in the same cookbook.

**Comment by someara** [comment]
That pattern is Doing It Wrong(tm). You're supposed to send the notification from the updated resource in the local run_context.

This is going to cause a world of confusion.
-s

**Comment by someara** [comment]
I've added an :immediate timing to the original notification and released it in 2.13.11.
It doesn't work quite like it used to, but it's better than nothing for now.

-s
